Problems Building on Visual Studio 2013

First off, something odd was happening when std::min and std::max were used. A #define appeared to be set for min and max, replacing them with nothing, but I was able to fix those errors by adding #undef min and #undef max at the start of the code in the relevant files.

Next, linker errors:

    Error	18	error LNK2019: unresolved external symbol "public: bool __cdecl encoder_picture_buffer::have_more_frames_to_encode(void)const " (?have_more_frames_to_encode@encoder_picture_buffer@@QEBA_NXZ) referenced in function en265_encode	C:\Users\____\Documents\GitHub\libde265\build\vc9-x86\en265.obj	dec265
    Error	19	error LNK2019: unresolved external symbol "public: void __cdecl encoder_picture_buffer::mark_image_is_outputted(int)" (?mark_image_is_outputted@encoder_picture_buffer@@QEAAXH@Z) referenced in function en265_free_packet	C:\Users\____\Documents\GitHub\libde265\build\vc9-x86\en265.obj	dec265
    Error	20	error LNK2019: unresolved external symbol "public: void __cdecl encoder_picture_buffer::release_input_image(int)" (?release_input_image@encoder_picture_buffer@@QEAAXH@Z) referenced in function en265_free_packet	C:\Users\____\Documents\GitHub\libde265\build\vc9-x86\en265.obj	dec265
    Error	21	error LNK2019: unresolved external symbol "public: __cdecl encoder_context::encoder_context(void)" (??0encoder_context@@QEAA@XZ) referenced in function en265_new_encoder	C:\Users\____\Documents\GitHub\libde265\build\vc9-x86\en265.obj	dec265
    Error	22	error LNK2019: unresolved external symbol "public: void __cdecl encoder_context::start_encoder(void)" (?start_encoder@encoder_context@@QEAAXXZ) referenced in function en265_start_encoder	C:\Users\____\Documents\GitHub\libde265\build\vc9-x86\en265.obj	dec265
    Error	23	error LNK2019: unresolved external symbol "public: enum de265_error __cdecl encoder_context::encode_picture_from_input_buffer(void)" (?encode_picture_from_input_buffer@encoder_context@@QEAA?AW4de265_error@@XZ) referenced in function en265_encode	C:\Users\____\Documents\GitHub\libde265\build\vc9-x86\en265.obj	dec265

--- a/.travis.yml
+++ b/.travis.yml
@@ -22,6 +22,7 @@ env:
   - HOST=i686-w64-mingw32 WINE=wine DECODESTREAMS=
   - HOST=x86_64-w64-mingw32 WINE=wine64 DECODESTREAMS=
   - HOST=arm-linux-gnueabihf WINE= DECODESTREAMS=
+  - HOST=cmake WINE= DECODESTREAMS=
 
 matrix:
   include:
@@ -42,10 +43,11 @@ install:
   - git clone https://github.com/strukturag/libde265-data.git
 
 before_script:
-  - ./autogen.sh
-  - if [ ! -z "$HOST" ]; then unset CC; fi
-  - if [ ! -z "$HOST" ]; then unset CXX; fi
-  - ./configure --host=$HOST
+  - if [ "$HOST" != "cmake" ]; then ./autogen.sh; fi
+  - if [ ! -z "$HOST" ] && [ "$HOST" != "cmake" ]; then unset CC; fi
+  - if [ ! -z "$HOST" ] && [ "$HOST" != "cmake" ]; then unset CXX; fi
+  - if [ "$HOST" != "cmake" ]; then ./configure --host=$HOST; fi
+  - if [ "$HOST" = "cmake" ]; then cmake .; fi
 
 script:
   - make

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,0 +1,65 @@
+project (libde265)
+cmake_minimum_required (VERSION 2.8)
+
+# The version number.
+set (NUMERIC_VERSION 0x01000000)
+set (PACKAGE_VERSION 1.0.0)
+
+include (${CMAKE_ROOT}/Modules/CheckCCompilerFlag.cmake)
+include (${CMAKE_ROOT}/Modules/CheckIncludeFile.cmake)
+include (${CMAKE_ROOT}/Modules/FindSDL.cmake)
+include (${CMAKE_ROOT}/Modules/FindThreads.cmake)
+
+CHECK_INCLUDE_FILE(malloc.h HAVE_MALLOC_H)
+CHECK_INCLUDE_FILE(stdint.h HAVE_STDINT_H)
+CHECK_INCLUDE_FILE(stdbool.h HAVE_STDBOOL_H)
+
+if (HAVE_MALLOC_H)
+  add_definitions(-DHAVE_MALLOC_H)
+endif()
+if (HAVE_STDINT_H)
+  add_definitions(-DHAVE_STDINT_H)
+endif()
+if (HAVE_STDBOOL_H)
+  add_definitions(-DHAVE_STDBOOL_H)
+endif()
+
+configure_file (
+  "${PROJECT_SOURCE_DIR}/libde265/de265-version.h.in"
+  "${PROJECT_BINARY_DIR}/libde265/de265-version.h"
+)
+
+if(CMAKE_COMPILER_IS_GNUCXX)
+  set(GCC 1)
+  add_definitions(-Wall)
+  set(CMAKE_CXX_FLAGS "-std=gnu++0x ${CMAKE_CXX_FLAGS}")
+endif()
+
+option(DISABLE_SSE "Disable SSE optimizations")
+if(NOT ${DISABLE_SSE} EQUAL OFF)
+  if(MSVC)
+    set(SUPPORTS_SSE4_1 1)
+  else()
+    CHECK_C_COMPILER_FLAG(-msse4.1 SUPPORTS_SSE4_1)
+  endif()
+endif()
+
+include_directories ("${PROJECT_SOURCE_DIR}")
+include_directories ("${PROJECT_BINARY_DIR}")
+include_directories ("${PROJECT_SOURCE_DIR}/libde265")
+if(MSVC)
+  include_directories ("${PROJECT_SOURCE_DIR}/extra")
+  add_definitions(-DHAVE_STDINT_H)
+  add_definitions(-DHAVE_STDBOOL_H)
+  add_definitions(-DNOMINMAX)
+endif()
+
+if(UNIX)
+  set(LIBDE265_LIBRARY_NAME de265)
+else()
+  set(LIBDE265_LIBRARY_NAME libde265)
+endif()
+
+add_subdirectory (libde265)
+add_subdirectory (dec265)
+add_subdirectory (enc265)

--- a/README.md
+++ b/README.md
@@ -89,6 +89,23 @@ Additional logging information can be turned on and off using these `./configure
 </pre>
 
 
+Build using cmake
+=================
+
+cmake scripts to build libde265 and the sample scripts `dec265` and `enc265` are
+included and can be compiled using these commands:
+
+```
+mkdir build
+cd build
+cmake ..
+make
+```
+
+See the [cmake documentation](http://www.cmake.org) for further information on
+using cmake on other platforms.
+
+
 Prebuilt binaries
 =================
 

--- a/build/vc9-x86/make-solutions.bat
+++ b/build/vc9-x86/make-solutions.bat
@@ -1,6 +1,0 @@
-@echo off
-::
-:: run this batch file to create a Visual Studion solution file for this project.
-:: See the cmake documentation for other generator targets
-::
-cmake -G "Visual Studio 9 2008" ..\..\dec265 && cmake-gui ..\..\dec265

--- a/dec265/CMakeLists.txt
+++ b/dec265/CMakeLists.txt
@@ -1,132 +1,41 @@
-set(CMAKE_CONFIGURATION_TYPES "Debug;Release;RelWithDebInfo" CACHE STRING "Configurations" FORCE)
-
-if(NOT CMAKE_BUILD_TYPE)
-    # default to Release build for GCC builds
-    set(CMAKE_BUILD_TYPE Release CACHE STRING
-        "Choose the type of build, options are: None(CMAKE_CXX_FLAGS or CMAKE_C_FLAGS used) Debug Release."
-        FORCE)
-endif()
-
-project (dec265)
-cmake_minimum_required (VERSION 2.8)
-
-SET(CMAKE_MODULE_PATH "${PROJECT_SOURCE_DIR}/../cmake" "${CMAKE_MODULE_PATH}")
-
-if ("${CMAKE_SIZEOF_VOID_P}" MATCHES 8)
-    set(X64 1)
-    add_definitions(-DX86_64)
-endif()
-
-# Enforce coding standards.  Full warnings and warnings as errors
-if(MSVC)
-    add_definitions(/W4 /D_CRT_SECURE_NO_WARNINGS)
-    add_definitions(/Ob2) # always inline
-    add_definitions(/Oi)  # enable intrinsics
-#    add_definitions(/MP)  # multithreaded build
-#    add_definitions(/TP)  # all files as cpp
-
-    include_directories(../extra)
-endif(MSVC)
-
-if("$ENV{CXX}" STREQUAL "icpc")
-    set(GCC 1)
-    add_definitions(-Wall -Wextra -Wunused-variable -Wunused-function -Wshadow -no-vec -msse4)
-elseif(CMAKE_COMPILER_IS_GNUCXX)
-    set(GCC 1)
-    add_definitions(-Wall -Wextra -Wunused-variable -Wunused-function -Wshadow -msse4)
-endif()
-
-option(USE_ASM "Use SSE41 SIMD optimize" ON)
-if(USE_ASM)
-    add_definitions(-DUSE_ASM=ASM_SSE4)
-    add_definitions(-DHAVE_SSE4_1)
-else(USE_ASM)
-    add_definitions(-DUSE_ASM=ASM_NONE)
-endif(USE_ASM)
-
-option(USE_VIDEOGFX "Use of VIDEOGFX" OFF)
-if(USE_VIDEOGFX)
-    add_definitions(-D HAVE_VIDEOGFX=1)
-else(USE_VIDEOGFX)
-    add_definitions(-D HAVE_VIDEOGFX=0)
-endif(USE_VIDEOGFX)
-
-option(BUILD_STATIC  "Static Lib" ON)
-if(BUILD_STATIC)
-    add_definitions(-DLIBDE265_STATIC_BUILD)
-endif(BUILD_STATIC)
-
-option(DE265_LOG_ERROR  "Enable DE265_LOG_ERROR" OFF)
-if(DE265_LOG_ERROR)
-    add_definitions(-DDE265_LOG_ERROR)
-endif(DE265_LOG_ERROR)
-
-option(DE265_LOG_INFO   "Enable DE265_LOG_INFO"  OFF)
-if(DE265_LOG_INFO)
-    add_definitions(-DDE265_LOG_INFO)
-endif(DE265_LOG_INFO)
-
-option(DE265_LOG_DEBUG  "Enable DE265_LOG_DEBUG" OFF)
-if(DE265_LOG_DEBUG)
-    add_definitions(-DDE265_LOG_DEBUG)
-endif(DE265_LOG_DEBUG)
-
-option(DE265_LOG_TRACE  "Enable DE265_LOG_TRACE" OFF)
-if(DE265_LOG_TRACE)
-    add_definitions(-DDE265_LOG_TRACE)
-endif(DE265_LOG_TRACE)
-
-include_directories(../ ../libde265)
-
-file(GLOB LIBSRC ../libde265/*.cc ../extra/*.c)
-file(GLOB LIBINC ../libde265/*.h ../extra/*.h)
-file(GLOB APPSRC dec265.cc)
-file(GLOB ASMSRC0 ../libde265/x86/sse.cc ../libde265/x86/sse-dct.cc)
-file(GLOB ASMSRC1 ../libde265/x86/sse-motion.cc)
-file(GLOB ASMINC ../libde265/x86/*.h)
-
-source_group(INC  FILES ${LIBINC})
-source_group(SRC  FILES ${LIBSRC})
-source_group(APP  FILES ${APPSRC})
-source_group(ASM  FILES ${ASMSRC0} ${ASMSRC1} ${ASMINC})
-
-if(USE_ASM)
-    SET(LIBSRC ${LIBSRC} ${ASMSRC0} ${ASMSRC1})
-    SET(LIBINC ${LIBINC} ${ASMINC})
-
-    # disable uninitialize check on VC, because x1 is not assign, remove later
-    #if(MSVC)
-    #    SET_SOURCE_FILES_PROPERTIES(${ASMSRC1} PROPERTIES COMPILE_FLAGS "/RTCs")
-    #endif(MSVC)
-endif(USE_ASM)
-
-if(UNIX)
-    SET(PLATFORM_LIBS pthread rt m)
-    SET_SOURCE_FILES_PROPERTIES(../extra/win32thread.c PROPERTIES HEADER_FILE_ONLY TRUE)
-endif(UNIX)
-
-# Main CLI application
-if(MSVC)
-    add_definitions(/wd4244) # type conversion, possible loss of data
-    add_definitions(/wd4100) # unreferenced formal parameter
-    add_definitions(/wd4505) # unreferenced local function has been removed
-    add_definitions(/wd4701) # potentially uninitialized local variable
-    add_definitions(/wd4127) # conditional expression is constant
-#    add_definitions(/wd4018) # signed/unsigned mismatch
-    add_definitions(/wd4189) # local variable is initialized but not referenced
-    add_definitions(/wd4715) # not all control paths return a value
-    add_definitions(/wd4324) # structure was padded due to __declspec(align())
-
-    set_source_files_properties(${LIBSRC} PROPERTIES LANGUAGE CXX)
-endif(MSVC)
-if(GCC)
-    add_definitions(-Wno-sign-compare)
-    add_definitions(-Wno-unused-parameter)
-endif(GCC)
-
-add_executable(dec265
-    ${APPSRC} ${EXTRAS} ${LIBSRC} ${LIBINC}
+set (dec265_sources
+  dec265.cc
 )
 
-target_link_libraries(dec265 ${PLATFORM_LIBS})
+set (hdrcopy_sources
+  hdrcopy.cc
+)
 
+if(MSVC)
+  set (dec265_sources
+    ${dec265_sources}
+    ../extra/getopt.c
+    ../extra/getopt_long.c
+  )
+  set (hdrcopy_sources
+    ${hdrcopy_sources}
+    ../extra/getopt.c
+    ../extra/getopt_long.c
+  )
+endif()
+
+if(SDL_FOUND)
+  add_definitions(-DHAVE_SDL)
+  include_directories ("${SDL_INCLUDE_DIR}")
+  set (dec265_sources
+    ${dec265_sources}
+    sdl.cc
+  )
+endif()
+
+add_executable (dec265 ${dec265_sources})
+
+target_link_libraries (dec265 ${LIBDE265_LIBRARY_NAME} ${SDL_LIBRARY})
+
+
+if(NOT MSVC)
+  # hdrcopy uses internal APIs that are not available when compiled for Windows
+  add_executable (hdrcopy ${hdrcopy_sources})
+
+  target_link_libraries (hdrcopy ${LIBDE265_LIBRARY_NAME})
+endif()

--- a/enc265/CMakeLists.txt
+++ b/enc265/CMakeLists.txt
@@ -1,0 +1,15 @@
+set (enc265_sources
+  enc265.cc
+)
+
+if(MSVC)
+  set (enc265_sources
+    ${enc265_sources}
+    ../extra/getopt.c
+    ../extra/getopt_long.c
+  )
+endif()
+
+add_executable (enc265 ${enc265_sources})
+
+target_link_libraries (enc265 ${LIBDE265_LIBRARY_NAME})

--- a/libde265/CMakeLists.txt
+++ b/libde265/CMakeLists.txt
@@ -1,0 +1,84 @@
+set (libde265_sources 
+  bitstream.cc
+  cabac.cc
+  de265.cc
+  deblock.cc
+  decctx.cc
+  nal-parser.cc
+  nal-parser.h
+  dpb.cc
+  dpb.h
+  image.cc
+  intrapred.cc
+  md5.cc
+  nal.cc
+  pps.cc
+  transform.cc
+  refpic.cc
+  sao.cc
+  scan.cc
+  sei.cc
+  slice.cc
+  sps.cc
+  util.cc
+  vps.cc
+  bitstream.h
+  cabac.h
+  deblock.h
+  decctx.h
+  image.h
+  intrapred.h
+  md5.h
+  nal.h
+  pps.h
+  transform.h
+  refpic.h
+  sao.h
+  scan.h
+  sei.h
+  slice.h
+  sps.h
+  util.h
+  vps.h
+  vui.h vui.cc
+  motion.cc motion.h
+  threads.cc threads.h
+  visualize.cc visualize.h
+  acceleration.h
+  fallback.cc fallback.h fallback-motion.cc fallback-motion.h
+  fallback-dct.h fallback-dct.cc
+  quality.cc quality.h
+  configparam.cc configparam.h
+  image-io.h image-io.cc
+  alloc_pool.h alloc_pool.cc
+  en265.h en265.cc
+  contextmodel.cc
+)
+
+if(MSVC)
+  set (libde265_sources
+    ${libde265_sources}
+    ../extra/win32cond.c
+    ../extra/win32cond.h
+  )
+endif()
+
+add_definitions(-DLIBDE265_EXPORTS)
+
+add_library(${LIBDE265_LIBRARY_NAME} SHARED ${libde265_sources})
+
+target_link_libraries(${LIBDE265_LIBRARY_NAME} ${CMAKE_THREAD_LIBS_INIT})
+
+if(CMAKE_SYSTEM_PROCESSOR STREQUAL "x86_64")
+  SET_TARGET_PROPERTIES(${LIBDE265_LIBRARY_NAME} PROPERTIES COMPILE_FLAGS "-fPIC")
+endif(CMAKE_SYSTEM_PROCESSOR STREQUAL "x86_64")
+
+add_subdirectory (encoder)
+
+target_link_libraries(${LIBDE265_LIBRARY_NAME} encoder)
+
+if(SUPPORTS_SSE4_1)
+  add_definitions(-DHAVE_SSE4_1)
+  add_subdirectory (x86)
+  target_link_libraries(${LIBDE265_LIBRARY_NAME} x86)
+endif()

--- a/libde265/encoder/CMakeLists.txt
+++ b/libde265/encoder/CMakeLists.txt
@@ -1,0 +1,18 @@
+set (encoder_sources 
+  analyze.cc analyze.h
+  encode.h encode.cc
+  encoder-params.h encoder-params.cc
+  encoder-context.h encoder-context.cc
+  encpicbuf.h encpicbuf.cc
+  sop.h sop.cc
+)
+
+add_library(encoder STATIC ${encoder_sources})
+
+if(CMAKE_SYSTEM_PROCESSOR STREQUAL "x86_64")
+  SET_TARGET_PROPERTIES(encoder PROPERTIES COMPILE_FLAGS "-fPIC")
+endif(CMAKE_SYSTEM_PROCESSOR STREQUAL "x86_64")
+
+add_subdirectory (algo)
+
+target_link_libraries(encoder algo)

--- a/libde265/encoder/algo/CMakeLists.txt
+++ b/libde265/encoder/algo/CMakeLists.txt
@@ -1,0 +1,20 @@
+set (algo_sources
+  algo.h algo.cc
+  coding-options.h coding-options.cc
+  ctb-qscale.h ctb-qscale.cc
+  cb-split.h cb-split.cc
+  cb-intrapartmode.h cb-intrapartmode.cc
+  cb-interpartmode.h cb-interpartmode.cc
+  cb-skip.h cb-skip.cc
+  cb-intra-inter.h cb-intra-inter.cc
+  cb-mergeindex.h cb-mergeindex.cc
+  tb-split.h tb-split.cc
+  tb-intrapredmode.h tb-intrapredmode.cc
+  pb-mv.h pb-mv.cc
+)
+
+add_library(algo STATIC ${algo_sources})
+
+if(CMAKE_SYSTEM_PROCESSOR STREQUAL "x86_64")
+  SET_TARGET_PROPERTIES(algo PROPERTIES COMPILE_FLAGS "-fPIC")
+endif(CMAKE_SYSTEM_PROCESSOR STREQUAL "x86_64")

--- a/libde265/x86/CMakeLists.txt
+++ b/libde265/x86/CMakeLists.txt
@@ -1,0 +1,24 @@
+set (x86_sources 
+  sse.cc sse.h
+)
+
+set (x86_sse_sources 
+  sse-motion.cc sse-motion.h sse-dct.h sse-dct.cc
+)
+
+add_library(x86 STATIC ${x86_sources})
+
+add_library(x86_sse STATIC ${x86_sse_sources})
+
+set(sse_flags "")
+
+if(NOT MSVC)
+  set(sse_flags "${sse_flags} -msse4.1")
+endif()
+
+target_link_libraries(x86 x86_sse)
+
+if(CMAKE_SYSTEM_PROCESSOR STREQUAL "x86_64")
+  SET_TARGET_PROPERTIES(x86 PROPERTIES COMPILE_FLAGS "-fPIC")
+  SET_TARGET_PROPERTIES(x86_sse PROPERTIES COMPILE_FLAGS "-fPIC ${sse_flags}")
+endif(CMAKE_SYSTEM_PROCESSOR STREQUAL "x86_64")

